### ISSUE TITLE
Create an abstract base class for the test functions

### ISSUE THREAD

**Issue by damar-wicaksono** [damar-wicaksono/uqtestfuns#4]
I'd like to implement the test functions from a unified abstract base class for a consistent interface.

I was thinking of the following abstract properties and methods:

- `spatial_dimension`: the dimension of the test function, i.e., the number of input variables
- `evaluate()`: the evaluation of the test function on a vector of inputs. It is assumed that the inputs are in the original domain (as presented, say, in the original paper).
- `transform_inputs()`: the transformation of a vector of inputs defined on a bounded domain (say, -1 to 1) to the original domain of the function.
- `__call__()`: calls `evaluate()` when an instance of the class is called with a vector of inputs

Each of the test functions will be implemented as a concrete class of this ABC. Example usage:

```python
import numpy as np
import uqtestfuns

# Create an instance of borehole test function
borehole = uqtestfuns.Borehole()

# Generate random sample of inputs
xx_prime = -1 + 2 * np.random.rand(1000,8)

# Transform the sampled inputs
xx = borehole.transform_inputs(xx_prime)

# Evaluate the inputs
yy = borehole(xx)

# or
yy = borehole.evaluate(xx)
```

Some possible extensions in the future:

- Store the information on the input parameters: the name, distribution, and parameters. _As what?_
- Ability to redefine the input parameters that changes how `evaluate()` and `transform_inputs()` compute things. _How to define the inputs? dict? list of dicts? data class?_ and _how to parse them?_

**Comment by damar-wicaksono** [comment]
The implementation has been included in the PR #15.

**Comment by damar-wicaksono** [comment]
Perhaps having an abstract class is not necessary as all the test functions implementations need not be a class. They can simply be instances of a single class.

**Comment by damar-wicaksono** [comment]
The new dataclass has been implemented in PR #17.

**Comment by damar-wicaksono** [comment]
This issue can be closed as the initial implementation according to PR #17 seems to be good enough.
